### PR TITLE
[FluidDyn] Optimize StokesTwoFluid build

### PR DESCRIPTION
**Pull request by vdivine**
Optimizing stokes two fluid element.

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp [hunk 1]
```diff
@@ -18,732 +18,750 @@ void Stokes3DTwoFluid::ComputeGaussPointLHSContribution(BoundedMatrix<double,16,
...
+        const double N_2 = N[2];
+        const double N_3 = N[3];
+
+        const double C_0_0 = C(0,0); const double C_0_1 = C(0,1); const double C_0_2 = C(0,2);
+        const double C_0_3 = C(0,3); const double C_0_4 = C(0,4); const double C_0_5 = C(0,5);
+        const double C_1_1 = C(1,1); const double C_1_2 = C(1,2); const double C_3_5 = C(3,5);
+        const double C_1_4 = C(1,4); const double C_1_5 = C(1,5); const double C_2_5 = C(2,5);
+        const double C_2_2 = C(2,2); const double C_2_3 = C(2,3); const double C_2_4 = C(2,4);
+        const double C_3_3 = C(3,3); const double C_4_4 = C(4,4); const double C_5_5 = C(5,5);
+        const double C_1_3 = C(1,3); const double C_3_4 = C(3,4); const double C_4_5 = C(4,5);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp]: Couldn't all these be references to avoid the allocation?

Comment (ddiezrod) [reply applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp]: Isnt it the same for doubles?

Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp]: You're right, it makes no difference in this case. Forget about my comment.

**Comment by rubenzorrilla** [comment]
By the way, it would be better to add this to the cpp template of the element and change the symbolic generator accordingly (see how it is done for the compressible NS element). Otherwise the symbolic generator will be no longer valid.

**Review by rubenzorrilla — DISMISSED** [review]
Fine for me. We should eventually do so with the other elements in the application.

**Comment by RiccardoRossi** [comment]
The entire point is that since
boost bounded_matrices have a variable size, then

C(4,3) implies something like    *(C.begin() + 4*row_size + 3) and the
multiplication needs to be performed AT RUNTIME.

experimentally i found out that this makes a quite big difference for the
optimizer

a  better solution here would be to use "mdspan" with a fixed size, which
would essentially mask everything with a fixed size so that multiplications
can be done at compiletime

https://github.com/kokkos/mdspan

**Comment by vdivine** [comment]
Thanks @rubenzorrilla 
I'm testing on some models to see if it decreases the elapsed time in build.

**Comment by RiccardoRossi** [comment]
@vdivine as it is it should provide a speedup already, however please note that there is space for a much bigger one by avoiding the use of heap allocated matrices in the element (particularly in the enrichment)

**Comment by vdivine** [comment]
@rubenzorrilla Thanks for your feedback.
I have also added some cpp tests for this element.
I can't merge as "Codacy Static Code Analysis" is failing, because variables in cpp template are not used (obviously). What could i do about that ?
@pooyan-dadvand

**Comment by vdivine** [comment]
@rubenzorrilla I added some unit tests, can you please approve again ? Thanks.

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp [hunk 2]
```diff
@@ -0,0 +1,236 @@
+//    |  /           |
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp:1]: I think you have something with your code editor. 4 spaces automatic indentation should be set.

File: applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp [hunk 3]
```diff
@@ -25,981 +25,996 @@ void Stokes3DTwoFluid::ComputeGaussPointLHSContribution(BoundedMatrix<double,16,
...
+lhs(15,6)=clhs362*clhs564 + clhs557;
+lhs(15,7)=clhs565;
+lhs(15,8)=clhs362*clhs676 + clhs610;
+lhs(15,9)=clhs362*clhs677 + clhs646;
+lhs(15,10)=clhs362*clhs678 + clhs675;
+lhs(15,11)=clhs679;
+lhs(15,12)=clhs363*clhs700;
+lhs(15,13)=clhs363*clhs712;
+lhs(15,14)=clhs363*clhs717;
+lhs(15,15)=clhs364*(clhs681 + clhs701 + clhs713);
```
Comment (RiccardoRossi) [inline applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp:1017]: note that there is another opportunity for opimization if you do something like

lhs(.,.) += weight*(...)

this was explored in some other PR

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp (hunk 2 above)
Comment (vdivine) [reply applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp:1]: OK Thanks, i will change that.

File: applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp (hunk 3 above)
Comment (vdivine) [reply applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp:1017]: Thanks Riccardo, i'll give a try.

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp (hunk 2 above)
Comment (vdivine) [reply applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp:1]: I removed tabs and add spaces, thanks.

File: applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp (hunk 3 above)
Comment (vdivine) [reply applications/FluidDynamicsApplication/custom_elements/stokes_3D_twofluid.cpp:1017]: Weight is not present in element_data, so it requires to add it and compute it. As it's a bigger development than just modifying template, i prefer doing it in next PR.

**Comment by vdivine** [comment]
@pooyan-dadvand

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+            KRATOS_CHECK_NEAR(RHS(6), 12.1078, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(7), 2866.56, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(8), 12.9595, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(9), 39.2958, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(10), 29.5256, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(11), 316699, 1.0);
+            KRATOS_CHECK_NEAR(RHS(12), 5.71452, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(13), 9.57769, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(14), 30.8809, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(15), -996.765, 1e-2);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp]: This increases a lot compilation time. Please use `KRATOS_CHECK_VECTOR_NEAR` instead.

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp [hunk 5]
```diff
@@ -0,0 +1,236 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Daniel Diez, Ruben Zorrilla
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp]:
```suggestion
//  License:            BSD License
//                            Kratos default license: kratos/license.txt
//
//  Main authors:    YOUR NAME IN HERE
```
Authorship (and minor spaces/tab issue).

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+//
+//
+
+// System includes
+#include <set>
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp]: Why do we need the `set` in here?

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+            KRATOS_CHECK_NEAR(RHS(6), 41.549, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(7), 411458, 1.0);
+            KRATOS_CHECK_NEAR(RHS(8), 19.7418, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(9), 32.0938, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(10), 44.015, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(11), 685764, 1.0);
+            KRATOS_CHECK_NEAR(RHS(12), 22.2078, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(13), 34.5488, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(14), 46.492, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(15), 950986, 1.0);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp]: `KRATOS_CHECK_VECTOR_NEAR`

File: applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+            KRATOS_CHECK_NEAR(RHS(6), -0.377754, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(7), 411.433, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(8), 0.0115203, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(9), 0.0211214, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(10), -0.378029, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(11), 685.739, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(12), 0.0112459, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(13), 0.0208464, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(14), -0.378303, 1e-2);
+            KRATOS_CHECK_NEAR(RHS(15), -8123.29, 1e-2);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/cpp_tests/test_stokes_two_fluid.cpp]: `KRATOS_CHECK_VECTOR_NEAR`

**Review by rubenzorrilla — COMMENTED** [review]
Aside of the minor comments in the test, this is in my opinion ready to go in.

**Comment by vdivine** [comment]
@rubenzorrilla Thanks for your review, i modified according to your suggestions.

**Review by rubenzorrilla — APPROVED** [review]
Fine if CI passes. Thanks!!!

**Comment by vdivine** [comment]
Thanks @rubenzorrilla

**Pull request merged**